apex/contrib/test/group_norm/test_group_norm.py

```diff
@@ -0,0 +1,153 @@
+#!/usr/bin/env python
+# coding: utf-8
+
+# Copyright (c) 2011-2023, NVIDIA CORPORATION.  All rights reserved.
+#
+# Redistribution and use in source and binary forms, with or without modification, are not permit-
+# ted.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR
+# IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND
+# FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL NVIDIA CORPORATION BE LIABLE
+# FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+# BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS;
+# OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
+# STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+import torch
+import unittest
+
+SKIP_TEST = None
+try:
+    from apex.contrib.group_norm.group_norm import cuda_group_norm_nhwc_one_pass
+    from apex.contrib.group_norm.group_norm import cuda_group_norm_nhwc_two_pass
+except ImportError as e:
+    SKIP_TEST = e
+
+
+# pytorch group norm requires same input type
+def torch_group_norm(x, g, w, b, eps, act=""):
+    xdtype, wdtype = x.dtype, w.dtype
+    if xdtype != wdtype:
+        x = x.to(dtype=wdtype)
+    y = torch.nn.functional.group_norm(x, g, w, b, eps)
+    if act in ["silu", "swish"]:
+        y = torch.nn.functional.silu(y)
+    if xdtype != wdtype and y.dtype != xdtype:
+        y = y.to(dtype=xdtype)
+    return y
+
+
+@unittest.skipIf(SKIP_TEST, f"{SKIP_TEST}")
+class GroupNormTest(unittest.TestCase):
+
+    def setUp(self, seed=0):
+        super().setUp()
+        torch.manual_seed(seed)
+
+    def verify_group_norm(self,
+                          tst_func,
+                          N=32,
+                          C=128,
+                          H=256,
+                          W=256,
+                          G=32,
```

Comment:
One last question: would we know the limitation of supported input shapes?
at glance, this is the only input shape tested

Reply:
N/H/W do not have lower (except >0) and upper bound limitations. Number of groups G is limited to 32. Supported number of channels C are: 128, 256, 320, 448, 512, 640, 768, 896, 960, 1024, 1280, 1344, 1536, 1792, 1920, 2048, 2240, 2560, 2688, 3072, 3136, 3584, 4096. Two pass algorithm might automatically support some other channels as well, but it is not tested. One pass algorithm supports only channels mentioned above.